test/e2e/controller/crash_test.go

```diff
@@ -58,6 +59,68 @@ func TestGameServerUnhealthyAfterDeletingPodWhileControllerDown(t *testing.T) {
 	assert.NoError(t, waitForAgonesControllerRunning(ctx))
 }
 
+func TestLeaderElectionAfterDeletingLeader(t *testing.T) {
+	logger := e2eframework.TestLogger(t)
+	gs := framework.DefaultGameServer(defaultNs)
+	ctx := context.Background()
+
+	// if !runtime.FeatureEnabled(runtime.FeatureSplitControllerAndExtensions) {
+	// 	t.Skip("Skip test. SplitControllerAndExtensions feature is not enabled")
+	// }
+
+	waitForAgonesControllerRunning(ctx)
+
+	list, err := getAgonesControllerPods(ctx)
+	require.NoError(t, err, "Could not get list of Extension pods")
+	if len(list.Items) == 1 {
+		t.Skip("Skip test. Leader Election is not enabled since there is only 1 controller")
+	}
+
+	replication := len(list.Items)
+
+	// Deleting one of the pods which cause the other one to most definitely be leader
+	willBeLeader := list.Items[0]
+	for i := range list.Items[1:] {
+		err = deleteAgonesControllerPod(ctx, list.Items[i])
+		require.NoError(t, err, "Could not delete controller pod")
+	}
+
+	err = deleteAgonesControllerPod(ctx, willBeLeader)
```

Comment:
It looks like you are:
* killing all replicas except the first
* then killing the first
* the waiting for everything to stabilize

Why? The way you've done it here, `willBeLeader` is pretty much irrelevant - you could have just killed all the pods.

The point of deleing the other pods is to shepherd the leader to `willBeLeader`. But you can't just kill it after doing that, otherwise you might as well have just restarted the whole replicaset.

What we talked about offoine was close to this, but not exactly:
* delete all but one as you do in line 82-86
* wait for all running (separately, just add a parameter to waitForAgonesControllerRunning if you want to ensure the count - you can add a wantReplicas parameter and then only check it if it's non-negative or something)
* then delete `willBeLeader`
* then proceed to verify the controller till works, without waiting again

the point is that shepherding it into one replica, waiting for the others to come back, then killing that one replica, is a pretty solid guarantee that you're killing the leader. The way you have it here also guarantees you killed the leader but it then waits for everything to come back up - we ideally want to show that right after the leader dies, we can still function. Make sense?

Reply:
Ooh I see I see. I think I have a better understanding now. I will try this.